File: swordOffer/minArray/minArray.cc
#include <iostream>
#include <vector>

using namespace std;

/*
// brace force
class Solution {
 public:
  int minArray(vector<int>& numbers) {

    if (numbers.empty()) {
      return 0;
    }

    int prev = numbers[0];
    for (int x : numbers) {
      if (x < prev) {
        return x;
      }
      prev = x;
    }

    return numbers.front();
  }
};
*/

class Solution {
 public:
  int minArray(vector<int>& numbers) {
    if (numbers.empty()) {
      return 0;
    }

    int left = 0;
    int right = numbers.size() - 1;
    int target = numbers.back();

    while (left < right) {
      int mid = (left + right) / 2;
      if (numbers[mid] > target) {
        left = mid + 1;
      } else if (numbers[mid] < target) {
        right = mid;
        target = numbers[right];
      } else {
        right--;
        target = numbers[right];
      }
    }

    return numbers[left];
  }
};

int main() {
  vector<int> input = {3, 4, 5, 1, 2};
  auto s = new Solution;
  cout << s->minArray(input);
  return 0;
}

File: swordOffer/go/arrStr/minArray.go
package main

func minArray(numbers []int) int {
	if len(numbers) == 0 {
		return 0
	}

	left, right := 0, len(numbers)-1
	// end := numbers[len(numbers) - 1] 要想清楚为什么每次比较的是numbers[right]而不是end？
	for left < right {
		mid := (left + right) / 2
		if numbers[mid] > numbers[right] {
			left = mid + 1
		} else if numbers[mid] < numbers[right] {
			right = mid
		} else {
			right-- // 数组是不减的，如果最后的几个数都一样
		}
	}
	if numbers[left] < numbers[right] {
		return numbers[left]
	}

	return numbers[right]
}
